tools/SwiftSyntax/SourceLength.swift

```diff
@@ -0,0 +1,87 @@
+//===------------------ SourceLength.swift - Source Length ----------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2018 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+/// The length a syntax node spans in the source code. From any AbsolutePosition
+/// you reach a node's end location by either adding its UTF-8 length or by
+/// inserting `lines` newlines and then moving `columns` columns to the right.
+public final class SourceLength {
+  public let lines: Int
+  public let columns: Int
+  public let utf8Length: Int
+
+  /// Construct the source length of a given text
+  public init(of text: String) {
+    var lines = 0
+    var columns = 0
+    var utf8Length = 0
+    for char in text {
+      let charLength = String(char).utf8.count
+      utf8Length += charLength
+      switch char {
+      case "\n", "\r\n":
+        lines += 1
+        columns = 0
+      default:
+        columns += charLength
+      }
+    }
+    self.lines = lines
+    self.columns = columns
+    self.utf8Length = utf8Length
+  }
+
+  public init(lines: Int, columns: Int, utf8Length: Int) {
+    self.lines = lines
+    self.columns = columns
+    self.utf8Length = utf8Length
+  }
+
+  /// A zero-length source length
+  public static let zero: SourceLength = SourceLength(lines: 0, columns: 0, utf8Length: 0)
+
+  /// Combine the length of two source length. Note that the addition is *not*
+  /// commutative (3 columns + 1 line = 1 line but 1 line + 3 columns = 1 line
+  /// and 3 columns)
+  public static func +(lhs: SourceLength, rhs: SourceLength) -> SourceLength {
+    let utf8Length = lhs.utf8Length + rhs.utf8Length
+    let lines = lhs.lines + rhs.lines
+    let columns: Int
+    if rhs.lines == 0 {
+      columns = lhs.columns + rhs.columns
+    } else {
+      columns = rhs.columns
+    }
+    return SourceLength(lines: lines, columns: columns, utf8Length: utf8Length)
+  }
+
+  public static func +=(lhs: inout SourceLength, rhs: SourceLength) {
+    lhs = lhs + rhs
+  }
+}
+
+/// Determine the AbsolutePosition by advancing the `lhs` by the given source
+/// length.
+public func +(lhs: AbsolutePosition, rhs: SourceLength) -> AbsolutePosition {
+  let utf8Offset = lhs.utf8Offset + rhs.utf8Length
+  let line = lhs.line + rhs.lines
+  let column: Int
+  if rhs.lines == 0 {
+    column = lhs.column + rhs.columns
+  } else {
+    column = rhs.columns + 1 // AbsolutePosition has 1-based columns
+  }
+  return AbsolutePosition(line: line, column: column, utf8Offset: utf8Offset)
+}
+
+public func +=(lhs: inout AbsolutePosition, rhs: SourceLength) {
+  lhs = lhs + rhs
```

Comment:
I don't feel comfortable using `inout` for class types. Have you tried `struct` for `AbsolutePosition` and `SourceLength`? I wonder how it does affect performance.

Reply:
We cannot make `AbsolutePosition` a struct since it is used inside `AtomicCache` and that requires a reference type. I'm trying to mimic `AbsolutePosition` being a value type as much as possible by making it immutable. The idea is that if we are able to use value types in `AtomicCache` (or decide to use a different caching strategy) we can just switch `AbsolutePosition` to be a struct with no further changes necessary.

Reply:
But we can box it to a class right? like:
```swift
final class Box<T> {
  init(_ value: T) { self.value = value }
  let var value: T
}
```
I'm OK to merge this PR without it. But I think it's worth to try it in follow-ups.

Reply:
I guess that's possible. I'll look into it in a follow-up PR.